### PR TITLE
GCLOUD: Support "private domains" plus many bugfixes

### PR DESCRIPTION
**Pull request by asn-iac**
Hi all,
First I'd like to acknowledge and thank the dnscontrol community for contributing to such a useful project.
We needed to implement some changes with the GCLOUD provider in a private fork of dnscontrol a few years back and finally have had time to submit some of these back to the public project.
This PR addresses a few existing bugs and adds some new features:
1) Simplify GCLOUD authentication logic and remove dependency on separate http.Client (looks related to #1409  https://github.com/StackExchange/dnscontrol/blob/31bf652e3409a15be649ab65d64382aa9df94553/providers/gcloud/gcloudProvider.go#L95-L96)
2) extend rate limit retry to additional API calls #2478 
3) address an issue executing changes after a new zone is created via 'dnscontrol push' #2479 
4) add support for the creation of zones with "private" visibility #2481 
5) implement batching of corrections to resolve 403 error when executing corrections on zones w/ >1000 changes #2480 

Happy to break these up into smaller PRs if needed.
TIA

### CODE REVIEW

**Comment by TomOnTime** [comment]
Wow!  What a generous contribution!

I'm on vacation now but will get to this when I return.

**Comment by TomOnTime** [comment]
@riyadhalnur could you please review?

File: providers/gcloud/gcloudProvider.go [hunk 1]
```diff
@@ -111,17 +112,56 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
 		oldRRsMap:     map[string]map[key]*gdns.ResourceRecordSet{},
 		zoneNameMap:   map[string]string{},
 	}
+	if len(metadata) != 0 {
+		err := json.Unmarshal(metadata, g)
+		if err != nil {
+			return nil, err
+		}
+		if len(g.Visibility) != 0 {
+			visibilityCheck := regexp.MustCompile("^(public|private)$")
```
Comment (riyadhalnur) [inline providers/gcloud/gcloudProvider.go:121]: Would prefer to compile this once at the top

File: providers/gcloud/gcloudProvider.go [hunk 2]
```diff
@@ -111,17 +112,56 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
...
+			if ok := visibilityCheck.MatchString(g.Visibility); !ok {
+				return nil, fmt.Errorf("GCLOUD :visibility set but does not match ^(public|private)$")
```
Comment (riyadhalnur) [inline providers/gcloud/gcloudProvider.go:123]: Maybe better to say `GCLOUD :visibility set but not one of public or private"`

File: providers/gcloud/gcloudProvider.go [hunk 3]
```diff
@@ -203,6 +243,20 @@ func (g *gcloudProvider) getZoneSets(domain string) (models.Records, error) {
 	return existingRecords, err
 }
 
+type msgs struct {
+	Additions, Deletions []string
+}
+
+type orderedChanges struct {
```
Comment (riyadhalnur) [inline providers/gcloud/gcloudProvider.go:250]: The struct below duplicates this. We can just have of them

**Review by riyadhalnur — CHANGES_REQUESTED** [review]
Thanks for these changes 😃. I'll test this branch against real data and get back soon

**Comment by asn-iac** [comment]
Thanks team!  I'm a bit tied up today but will review comments early next week

File: providers/gcloud/gcloudProvider.go [hunk 4]
```diff
@@ -73,28 +78,24 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
 	// fix it if we find that.
 
 	ctx := context.Background()
-	var hc *http.Client
+	var opt option.ClientOption
 	if key, ok := cfg["private_key"]; ok {
 		cfg["private_key"] = strings.Replace(key, "\\n", "\n", -1)
+		printer.Debugf("GCLOUD :private_key configured, attempting key-based authentication\n")
```
[the remark this answers is not in the document]
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: removed

File: providers/gcloud/gcloudProvider.go [hunk 5]
```diff
@@ -73,28 +78,24 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
...
 		}
-		hc = config.Client(ctx)
+		opt = option.WithTokenSource(config.TokenSource(ctx))
 	} else {
-		var err error
-		hc, err = gauth.DefaultClient(ctx, "https://www.googleapis.com/auth/ndev.clouddns.readwrite")
-		if err != nil {
-			return nil, fmt.Errorf("no creds.json private_key found and ADC failed with:\n%s", err)
-		}
+		printer.Debugf("GCLOUD :using application default credentials\n")
```
[the remark this answers is not in the document]
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: removed

File: providers/gcloud/gcloudProvider.go [hunk 6]
```diff
@@ -111,17 +112,56 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
...
+		}
+		if len(g.Visibility) != 0 {
+			visibilityCheck := regexp.MustCompile("^(public|private)$")
+			if ok := visibilityCheck.MatchString(g.Visibility); !ok {
+				return nil, fmt.Errorf("GCLOUD :visibility set but does not match ^(public|private)$")
+			}
+			printer.Printf("GCLOUD :visibility %s configured\n", g.Visibility)
+		}
+		for i, v := range g.Networks {
+			networkURLCheck := regexp.MustCompile("^" + selfLinkBasePath + "[a-z][-a-z0-9]{4,28}[a-z0-9]/global/networks/[a-z]([-a-z0-9]{0,61}[a-z0-9])?$")
```
[the remark this answers is not in the document]
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: moved all of the regexp.MustComplie() calls to var block at the top - hope that was what you were looking for.  let me know if you had something else in mind.

File: providers/gcloud/gcloudProvider.go [hunk 7]
```diff
@@ -111,17 +112,56 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
...
+			if networkURLCheck.MatchString(v) {
+				// the user specified a fully qualified network url
+			} else {
+				networkNameCheck := regexp.MustCompile("^[a-z]([-a-z0-9]{0,61}[a-z0-9])?$")
```
[the remark this answers is not in the document]
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: moved all of the regexp.MustComplie() calls to var block at the top - hope that was what you were looking for.  let me know if you had something else in mind.

File: providers/gcloud/gcloudProvider.go [hunk 8]
```diff
@@ -111,17 +112,56 @@ func New(cfg map[string]string, metadata json.RawMessage) (providers.DNSServiceP
...
+		if len(g.Visibility) != 0 {
+			visibilityCheck := regexp.MustCompile("^(public|private)$")
+			if ok := visibilityCheck.MatchString(g.Visibility); !ok {
+				return nil, fmt.Errorf("GCLOUD :visibility set but does not match ^(public|private)$")
+			}
+			printer.Printf("GCLOUD :visibility %s configured\n", g.Visibility)
+		}
+		for i, v := range g.Networks {
+			networkURLCheck := regexp.MustCompile("^" + selfLinkBasePath + "[a-z][-a-z0-9]{4,28}[a-z0-9]/global/networks/[a-z]([-a-z0-9]{0,61}[a-z0-9])?$")
+			if networkURLCheck.MatchString(v) {
```
[the remark this answers is not in the document]
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: updated logic to remove if-else

File: providers/gcloud/gcloudProvider.go (hunk 1 above)
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: moved all of the regexp.MustComplie() calls to var block at the top - hope that was what you were looking for.  let me know if you had something else in mind.

File: providers/gcloud/gcloudProvider.go (hunk 2 above)
Comment (asn-iac) [reply providers/gcloud/gcloudProvider.go]: updated message

**Pull request merged**